### ISSUE TITLE
[ESP32-S3] Compiler optimizations break ULP-RISCV code in Light Sleep mode (IDFGH-11652)

### ISSUE THREAD

**Issue by kutukvpavel** [espressif/esp-idf#12765]
UPD: updated the issue name and description based on my new findings.

### Answers checklist.

- [X] I have read the documentation [ESP-IDF Programming Guide](https://docs.espressif.com/projects/esp-idf/en/latest/) and the issue is not addressed there.
- [X] I have updated my IDF branch (master or release) to the latest version and checked that the issue is present there.
- [X] I have searched the issue tracker for a similar issue and not found a similar issue.

### IDF version.

v5.1.2

### Espressif SoC revision.

ESP32-S3 rev 0.1 (ESP32-S3-WROOM-1 MON4R2, P1N8R8)

### Operating System used.

Linux

### How did you build your project?

VS Code IDE

### If you are using Windows, please specify command line type.

None

### Development Kit.

Custom board, ESP32-S3-DevKitC-1 v1.6

### Power Supply used.

External 3.3V

### What is the expected behavior?

ULP-RISCV and Sleep-related code should work regardless of compiler optimization level.

### What is the actual behavior?

Increasing optimization level past `-O0` progressively breaks ULP functionality in Light Sleep. Deep Sleep doesn't seem to be affected.

#### Example 1: `-Og` + ADC under ULP-RISCV control
ADC1 works fine w/ULP-RISCV only in following "sleep mode + RTC clock source" combinations:
- Deep sleep + External XTAL_32K
- Deep sleep + Internal RC-oscillator 136kHz
- Light sleep + Internal RC-oscillator 136kHz

It fails when "Light Sleep + External XTAL_32K" combination is used. In this case, ADC1 samples input channel only once. Supposedly, it does so during the first `ulp_riscv_adc_read_channel` call since boot, and any subsequent `ulp_riscv_adc_read_channel` calls return the exact same value, no matter what the actual voltage on the input pin is.

#### Example 2: `-O2` or `-Os` + ULP wakeup based on ADC reading
Once again, if "Light Sleep + External XTAL_32K" combination is used, ULP fails to wake the SoC up at all. See the example code below.

### Steps to reproduce.
1. Equip any ESP32-S3 board with an external 32.768kHz crystal with appropriate loading capacitors (XTAL_32K).
2. Flash the following adapted ulp_riscv_adc example code (adapted to perform periodic ADC measurement output to debug monitor), with "RTC clock = Internal RC 136kHz" as well as "App optimization level = `Og`" set in menuconfig:

**ulp_riscv_adc_example_main.c**
```
#include <stdio.h>
#include <inttypes.h>
#include "esp_sleep.h"
#include "ulp_riscv.h"
#include "ulp_adc.h"
#include "esp_adc/adc_oneshot.h"
#include "ulp_main.h"
#include "ulp/example_config.h"

extern const uint8_t ulp_main_bin_start[] asm("_binary_ulp_main_bin_start");
extern const uint8_t ulp_main_bin_end[]   asm("_binary_ulp_main_bin_end");

static adc_oneshot_unit_handle_t adc_handle;

static void init_ulp_program(void);

void app_main(void)
{
    ESP_ERROR_CHECK(esp_sleep_pd_config(ESP_PD_DOMAIN_RTC_PERIPH, ESP_PD_OPTION_ON)); //So that ULP can use SAR ADC
    //ESP_ERROR_CHECK(esp_sleep_pd_config(ESP_PD_DOMAIN_RC_FAST, ESP_PD_OPTION_ON));
    init_ulp_program();
    ESP_ERROR_CHECK(esp_sleep_enable_ulp_wakeup());
    while (1)
    {
        ESP_ERROR_CHECK(ulp_riscv_run());
        esp_light_sleep_start();
        printf("ULP-RISC-V woke up the main CPU\n");
        printf("Value = %"PRIu32"\n", ulp_wakeup_result);
    }
}

static void init_ulp_program(void)
{
    ulp_adc_cfg_t cfg = {
        .adc_n    = EXAMPLE_ADC_UNIT,
        .channel  = EXAMPLE_ADC_CHANNEL,
        .width    = EXAMPLE_ADC_WIDTH,
        .atten    = EXAMPLE_ADC_ATTEN,
        .ulp_mode = ADC_ULP_MODE_RISCV,
    };

    ESP_ERROR_CHECK(ulp_adc_init(&cfg));

    esp_err_t err = ulp_riscv_load_binary(ulp_main_bin_start, (ulp_main_bin_end - ulp_main_bin_start));
    ESP_ERROR_CHECK(err);

    ulp_set_wakeup_period(0, 20000);
}
```

**main.c (ULP program)**
```
#include <stdint.h>
#include "ulp_riscv_utils.h"
#include "ulp_riscv_adc_ulp_core.h"

#include "example_config.h"

#define ULP_TIMER_PERIOD 20u //mS
#define WAKEUP_PERIOD 1000u //mS

volatile int32_t wakeup_result;

int main (void)
{
    static uint32_t cycles = 0;

    wakeup_result = ulp_riscv_adc_read_channel(EXAMPLE_ADC_UNIT, EXAMPLE_ADC_CHANNEL);

    if (++cycles > (WAKEUP_PERIOD / ULP_TIMER_PERIOD))
    {
        cycles = 0;
        ulp_riscv_wakeup_main_processor();
    }

    return 0;
}
```

**example_config.h**
Stays the same.

3. Observe normal operation (ADC readings correspond to input voltage, albeit with some delay).
4. In menuconfig set "RTC clock = External XTAL_32K", build and flash.
5. Observe the output listed below:
```plain
<Bootloader log omitted>
entry 0x403c9894
ULP-RISC-V woke up the main CPU
Value = 1867
ULP-RISC-V woke up the main CPU
Value = 1867
<Further output remains the same despite any changes of the input voltage>
```
6. Set app optimization level to `Os` in menuconfig, build, flash. Observe that the SoC does not wake up at all now.

### More Information.

See comment below.

**Comment by kutukvpavel** [comment]
Now, here's something interesting. If I set Debug Log Verbosity to "Debug" (used to be "Warning") in menuconfig the problem goes away.

**Comment by kutukvpavel** [comment]
Well, I have no idea why, for example, placing `periph_ll_get_rst_en_mask(...)` in `pragma GCC optimize ("O0")` helps with this issue, or why optimization level influences program behavior in this case at all.
However, the root cause seems to be the same as in #12426. Moving discussion over there.

**Comment by KonstantinKondrashov** [comment]
Hi @kutukvpavel!
I want to clarify that you use the `External 32 kHz crystal`, not `External 32 kHz oscillator at XTAL_32K_P pin` option. And the 32 kHz crystal is connected to the right pins (XTAL_32K_P and XTAL_32K_N).
https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/api-reference/system/system_time.html#rtc-timer-clock-sources

You have not mentioned which example you used, probably based on this one [ulp_riscv adc](https://github.com/espressif/esp-idf/tree/master/examples/system/ulp/ulp_riscv/adc).

Do you observe the same issue on the master too? Could you try it?
I do not have a 32k Xtal near me for now to test it, I will ask somebody to try it.

You closed this issue, why? Is your guess about the relation to the optimization option not confirmed?

**Comment by kutukvpavel** [comment]
@KonstantinKondrashov Sorry for the delay, a little busy ATM.

First of all, I indeed use an external crystal (resonator), not an oscillator (both in menuconfig and in hardware). It is connected as per Hardware Design Guidelines.

As for the example, I started with the one you correctly deduced, then modified it as describe above, for it to use Light Sleep instead of Deep Sleep (but keep as much initialization code as possible the same), because the issue only affects Light Sleep (for me).

I'm afraid I did not test master, but I'll try to do so ASAP.

I closed this one as a duplicate, since A) the solution proposed over at #12426 solved this issue for me and B) I tried to disable all optimizations in menuconfig for the example provided over at #12426 and it also helped. Therefore I concluded that my issue is just a different manifestation of the same bug under slightly different conditions. Which is expected for UB- and optmization-related issues.
